### PR TITLE
Fix/legend resize

### PR DESCRIPTION
**Pull request by timelyportfolio**
This is my first attempt at addressing #348 and https://github.com/ropensci/plotly/issues/525.  In summary, I

- moved some code out of the `if(firstRender==true` block as suggested in [@etpinard issue comment](https://github.com/plotly/plotly.js/issues/348#issuecomment-200411524)
- fixed `clipPath` enter,append,exit pattern to update on resize
- added `scrollheight` argument to the `scrollHandler` and remove `eventListeners` on resize to avoid scope problems
- changed drag behavior to `d3.behavior.drag()` which should properly handle touch and allow to work properly in RStudio viewer

Since this is my first pull, I apologize if I missed any of the contribution guidelines.  Also, other than a visual test, I was not sure how to test.  I would appreciate some advice here.

### CODE REVIEW

**Comment by timelyportfolio** [comment]
mistake #1 forgot to lint.  doing now.

File: src/components/legend/draw.js [hunk 1]
```diff
@@ -229,44 +231,46 @@ module.exports = function draw(gd) {
...
-                    scrollHandler(e.movementY);
-                }
-            }
+        scrollHandler(0,scrollheight);
 
-            function mUp() {
-                scrollBar.node().removeEventListener('mousemove', mMove);
-                window.removeEventListener('mouseup', mUp);
-            }
+        legend.on('wheel',null);
```
Comment (mdtusz) [inline src/components/legend/draw.js]: Does this need to be called with `null` before the handler?

File: src/components/legend/draw.js [hunk 2]
```diff
@@ -229,44 +231,46 @@ module.exports = function draw(gd) {
...
     }
 
-    function scrollHandler(delta) {
+
+    function scrollHandler(delta, scrollheight) {
 
         var scrollBarTrack = scrollheight - constants.scrollBarHeight - 2 * constants.scrollBarMargin,
             translateY = scrollBox.attr('data-scroll'),
-            scrollBoxY = Lib.constrain(translateY - delta, Math.min(scrollheight - opts.height, 0), 0),
+            scrollBoxY = Lib.constrain(translateY - delta, scrollheight-opts.height, 0),
```
Comment (mdtusz) [inline src/components/legend/draw.js]: :cow2: spacing between `scrollheight` and `opts.height`.

**Comment by mdtusz** [comment]
This looks pretty good to me. We'll wait for the all-clear from @etpinard then merge it in!

File: src/components/legend/draw.js (hunk 2 above)
Comment (timelyportfolio) [inline src/components/legend/draw.js]: oh thanks.  should have thought about using `d3` wheel also.  great change!

**Comment by etpinard** [comment]
Yep. I like this. I was worried about using d3 events looking forward to d3's 4.0 release. But looks like events on selection will be part of the stock [d3-selection](https://github.com/d3/d3-selection#handling-events) package, which we'll obviously need. So :beers: great solution!

@timelyportfolio Can you add a test case in `legend_scroll_test.js` similar to this [case](https://github.com/plotly/plotly.js/blob/bbb8205903cdd34ae019d5cd6a2c197e9a2550e7/test/jasmine/tests/legend_scroll_test.js#L80)? Your test case should feature a call to `Plotly.relayout` with a layout `height/width` update to check that the legend dimensions update properly and that only one handler of each type is attached to the legend nodes.

**Comment by etpinard** [comment]
It would be nice to merge this one in for the 1.7.1 release at some point this week.

@timelyportfolio, are you going to have time to work on adding a test case in the next few days?

**Comment by timelyportfolio** [comment]
@etpinard I should be able to wrap this up today.

**Comment by etpinard** [comment]
@timelyportfolio can you also check that this [example](http://community.plot.ly/t/default-showlegend-changed-with-update/708/2) works as expected? Thanks!

**Comment by timelyportfolio** [comment]
@etpinard yes, will be happy to check this also

File: test/jasmine/tests/legend_scroll_test.js [hunk 3]
```diff
@@ -88,10 +92,28 @@ describe('The legend', function() {
...
+            Plotly.relayout(gd, {'height':gd._fullLayout.height/2}).then(function() {
+                var legend = document.getElementsByClassName('legend')[0];
+                var legendHeight = getBBox(legend).height;
+                //legend still exists and not duplicated
+                expect(countLegendGroups(gd)).toBe(1);
+                expect(countLegendClipPaths(gd)).toBe(1);
+                
+                // clippath resized to new height less than new plot height
+                expect(+legendHeight).toBe(getPlotHeight(gd));
+                expect(+legendHeight).toBeLessThan(+origLegendHeight);
```
Comment (mdtusz) [inline test/jasmine/tests/legend_scroll_test.js]: Could you check against the computed height using `toBeCloseTo`?

It helps catch potential bugs earlier than just relative size comparison.

File: test/jasmine/tests/legend_scroll_test.js [hunk 4]
```diff
@@ -122,6 +146,17 @@ describe('The legend', function() {
                 done();
             });
         });
+        
+        it('should resize when traces added', function(done) {
+            var origLegendHeight = getBBox(legend).height;
+            
+            Plotly.addTraces(gd, { x: [1,2,3], y: [4,3,2], name: 'Test2' }).then(function() {
+                // clippath resized to show new trace
+                expect(+legendHeight).toBeGreaterThan(+origLegendHeight);
```
Comment (mdtusz) [inline test/jasmine/tests/legend_scroll_test.js]: Same here - check `toBeCloseTo(theValue, somePrecision)`

File: test/jasmine/tests/legend_scroll_test.js (hunk 3 above)
Comment (timelyportfolio) [inline test/jasmine/tests/legend_scroll_test.js]: sure, will be happy to; sorry for the flurry of commits; my local set up broke this morning.

**Comment by timelyportfolio** [comment]
I'm having a very difficult time debugging these simple tests with testing broken on my devel machine.  I don't want to hold up the 1.7.1 release.  Please let me know if you help me see the obvious that I am missing. Thanks. Will try again tomorrow.

**Comment by etpinard** [comment]
@timelyportfolio I pulled down your branch on added a test fixup commit in https://github.com/plotly/plotly.js/pull/365.

This issue is now a week old, and quite a few users have noticed its effects, so I decided to get your patch in ASAP. Thanks for work :beers:

**Comment by timelyportfolio** [comment]
thanks!

**Pull request closed without merging**